File: src/data/courses/dsa/two-pointers-sliding-window/sliding-window-fixed-size/sliding-window-maximum/code/python/sliding_window_maximum.py
# ============================================================
# Sliding Window Maximum
# ============================================================

from typing import List
from collections import deque

# ============================================================
# Solution 1: Brute Force
# Time: O(N×K) | Space: O(1)
# ============================================================
class Solution1:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        return [max(nums[i:i+k]) for i in range(len(nums) - k + 1)]

# ============================================================
# Solution 2: Monotonic Decreasing Deque (Optimal)
# Time: O(N) | Space: O(K)
# ============================================================
class Solution2:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        dq = deque()  # Stores indices; front = current max
        result = []

        for i in range(len(nums)):
            # Remove expired indices from front
            while dq and dq[0] < i - k + 1:
                dq.popleft()

            # Remove smaller elements from back
            while dq and nums[dq[-1]] <= nums[i]:
                dq.pop()

            dq.append(i)

            # Window is full — front is the max
            if i >= k - 1:
                result.append(nums[dq[0]])

        return result

File: src/data/courses/dsa/two-pointers-sliding-window/sliding-window-fixed-size/sliding-window-maximum/code/cpp/SlidingWindowMaximum.cpp
#include <vector>
#include <deque>
#include <algorithm>
using namespace std;
// ============================================================
// Solution 1: Brute Force — scan each window for max — O(N*K)
// ============================================================
class Solution1 {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        int n = nums.size();
        vector<int> result;
        for (int i = 0; i <= n - k; i++) {
            int mx = nums[i];
            for (int j = i + 1; j < i + k; j++) mx = max(mx, nums[j]);
            result.push_back(mx);
        }
        return result;
    }
};

// ============================================================
// Solution 2: Monotonic Decreasing Deque — O(N) Time
// ============================================================
class Solution2 {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        int n = nums.size();
        deque<int> dq; // Indices in decreasing value order
        vector<int> result;
        for (int i = 0; i < n; i++) {
            while (!dq.empty() && dq.front() <= i - k) dq.pop_front(); // Expire
            while (!dq.empty() && nums[dq.back()] <= nums[i]) dq.pop_back(); // Remove smaller
            dq.push_back(i);
            if (i >= k - 1) result.push_back(nums[dq.front()]); // Front = max
        }
        return result;
    }
};
